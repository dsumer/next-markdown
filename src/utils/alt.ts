/**
 * If the alt is a JSON, the JSON key/value will be transfered to `node.properties`.
 *
 * For example, this markdown:
 *
 * ```md
 * ![{"alt":"20211029 - edit bot demo","poster":"https://frouo.com/poster.jpg"}](https://frouo.com/video.mp4)
 * ```
 *
 * will generate:
 *
 * ```html
 * <video src="https://frouo.com/video.mp4" alt="20211029 - edit bot demo" width="100%" controls poster="https://frouo.com/poster.jpg"></video>
 * ```
 */
export const extractDataFromAlt = (alt: string) => {
  try {
    return JSON.parse(alt) as { [key: string]: any };
  } catch (error) {
    return alt;
  }
};