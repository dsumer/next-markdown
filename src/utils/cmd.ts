import { exec } from 'child_process';

export const cmd = (commandLine: string) => {
  return new Promise<string>((resolve, reject) => {
    exec(commandLine, (error, stdout, stderr) => {
      if (error) {
        reject(error);
      }
      resolve(stdout || stderr);
    });
  });
};
