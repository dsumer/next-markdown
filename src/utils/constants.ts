import { join } from 'path';

export const pathToLocalGitRepo = join(process.cwd(), '.git/next-md/');
export const pathToNextmdLastUpdate = join(process.cwd(), '.git/next-md-last-update');
export const pathToNextmdBranch = join(process.cwd(), '.git/next-md-branch');
