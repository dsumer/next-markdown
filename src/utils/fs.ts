import fs from 'fs';
import { join, parse, sep } from 'path';
import { File, TreeObject } from '../types';

export const pathToContent = (userPathToContent: string, remote: boolean) =>
  join(remote ? './.git/next-md' : './', userPathToContent);

export const exclude = (object: TreeObject) => {
  if (object.type === 'file' && object.name.endsWith('.md') === false && object.name.endsWith('.mdx') === false) {
    return true;
  }

  if (object.name === '.git') {
    return true;
  }

  if (object.type === 'dir' && object.children.length === 0) {
    return true;
  }

  return false;
};

export const treeSync = (path: string, absoluteToRelative: (path: string) => string) => {
  const res = fs
    .readdirSync(path)
    .map((e) => ({ name: e, dir: path }))
    .map(
      (
        (fileSystem: typeof fs) =>
        (e): TreeObject | null => {
          const _path = join(e.dir, e.name);
          const isDir = fileSystem.lstatSync(_path).isDirectory();
          if (e.name === '.git') {
            return null; // no need to inspect recursively this
          } else {
            return isDir
              ? {
                  type: 'dir',
                  name: e.name,
                  path: absoluteToRelative(_path),
                  children: treeSync(_path, absoluteToRelative),
                }
              : {
                  type: 'file',
                  name: e.name,
                  path: absoluteToRelative(_path),
                };
          }
        }
      )(fs),
    )
    .flatMap((f) => (f ? [f] : []))
    .filter((e) => exclude(e) === false);

  return res;
};

export const flatFiles = (tree: TreeObject[]): File[] => {
  const flatRecursively = (object: TreeObject): TreeObject[] =>
    object.type === 'dir' ? object.children.flatMap(flatRecursively) : [object];

  return tree.flatMap(flatRecursively) as File[];
};

export const generateNextmd = (relativeFilePath: string) => {
  return relativeFilePath
    .split(sep)
    .map((e) => e.replace(/^_(\s+)?/, '')) // replace the first occurence of "_" at the beginning of the string with ""
    .map((e) => e.replace(/^\[(.*?)\](\s+)?_?/, '')) // replace the first occurence of "[whatever-you-want]+whitespaces(if-any)" at the beginning of string with ""
    .map((e, index, arr) => {
      if (index === arr.length - 1) {
        const lastElement = e;
        const parsed = parse(lastElement);
        if (parsed.name === 'index') {
          return undefined;
        } else {
          return parsed.name;
        }
      } else {
        return e;
      }
    })
    .map((e) => e?.trim())
    .flatMap((e) => (e ? [e] : []));
};

export const getPostFilesFromNextmd = async (files: File[], nextmd: string[]) =>
  files.filter((e) => parse(e.path).dir === nextmd.join('/'));

export const isMDX = (filePath: string) => filePath.endsWith('.mdx');

export const readFileSyncUTF8 = (filePath: string) => fs.readFileSync(filePath).toString('utf-8');

export const isDraft = (relativeFilePath: string) =>
  relativeFilePath.split(sep).find((e) => e.startsWith('_')) !== undefined;
