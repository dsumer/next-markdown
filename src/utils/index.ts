export * from './alt';
export * from './fs';
export * from './git';
export * from './markdown';
export * from './logger';
