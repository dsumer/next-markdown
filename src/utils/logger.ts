export const consoleLogNextmd = (...args: (string | undefined | null)[]) => {
  if (process.env.NODE_ENV === 'test') {
    return;
  }

  args.unshift('[nextmd]');
  console.log.apply(this, args); // tslint:disable-line:no-console
};
