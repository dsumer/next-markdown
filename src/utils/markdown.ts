import fs from 'fs';
import matter from 'gray-matter';
import { Root } from 'mdast';
import { join } from 'path';
import rehypeSlug from 'rehype-slug';
import rehypeStringify from 'rehype-stringify';
import remarkParse from 'remark-parse';
import remarkRehype from 'remark-rehype';
import { unified } from 'unified';
import { visit } from 'unist-util-visit';
import { File, YAMLFrontMatter } from '../types';
import { extractDataFromAlt } from './alt';
import { getNextmdFromFilePath, isMDX } from './fs';
import { getTableOfContents } from './table-of-contents';

export const getPostsFromNextmd = async <T extends YAMLFrontMatter>(
  files: File[],
  localRepoPath: string,
  nextmd: string[],
) => {
  type PostFile = { file: File; date: string };

  const posts = files.reduce<PostFile[]>((prev, curr) => {
    const matches = curr.name.match(/^\d{4}-\d{2}-\d{2}/i);
    if (curr.path.startsWith(join(localRepoPath, nextmd.join('/'))) && matches && matches.length > 0) {
      return prev.concat([{ file: curr, date: matches[0] }]);
    }
    return prev;
  }, []);

  return posts.length === 0
    ? null
    : await Promise.all(
        posts.map(async (e) => {
          const postPageData = await readMarkdownFile<T>(e.file.path);
          const postNextmd = getNextmdFromFilePath(e.file.path, localRepoPath);
          return {
            file: e.file,
            data: {
              ...postPageData,
              nextmd: postNextmd,
              date: e.date,
            },
          };
        }),
      );
};

export const readMarkdownFile = async <T extends YAMLFrontMatter>(filePath: string) => {
  const importMdxSerialize = async () => (await import('next-mdx-remote/serialize')).serialize;

  const rawdata = fs.readFileSync(filePath).toString('utf-8');
  const { frontMatter, content } = extractFrontMatter<T>(rawdata);
  const mdx = isMDX(filePath);

  return {
    frontMatter,
    html: mdx ? null : await markdownToHtml(content),
    mdxSource: mdx ? await importMdxSerialize().then((serialize) => serialize(content)) : null,
    tableOfContents: getTableOfContents(content),
  };
};

export const extractFrontMatter = <T extends YAMLFrontMatter>(rawdata: string) => {
  const { data, content } = matter(rawdata);

  return {
    frontMatter: data as T,
    content,
  };
};

export const markdownToHtml = async (markdown: string) => {
  const result = await unified()
    .use(remarkParse)
    .use(remarkRehype)
    .use(rehypeVideos)
    .use(rehypeStringify)
    .use(rehypeSlug)
    .process(markdown);

  return String(result);
};

export const rehypeVideos = (): ((tree: Root) => void) => {
  return (tree) => {
    visit(tree, 'element', (node: any) => {
      if (node.tagName === 'img' && node.properties && typeof node.properties.src === 'string') {
        const url = node.properties.src;

        if (url.includes('.mov') || url.includes('mp4')) {
          const altData = extractDataFromAlt(node.properties.alt);

          node.tagName = 'video';
          node.properties.width = '100%';
          node.properties.controls = true;

          if (typeof altData === 'string') {
            node.properties.alt = altData;
          } else {
            Object.entries(altData).forEach(([key, value]) => {
              node.properties[key] = value;
            });
          }
        }
      }
    });
  };
};
