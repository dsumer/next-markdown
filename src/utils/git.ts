import fs from 'fs';
import { cmd } from './cmd';
import { treeSync } from './fs';
import { consoleLogNextmd } from './logger';

const pathToNextmdLastUpdate = '.git/next-md-last-update';
const pathToNextmdBranch = '.git/next-md-branch';

type GitRepo = {
  remoteUrl: string;
  branch: string;
};

/**
 * @param pathToContent Absolute path to content
 */
export const treeContentRepo = async (
  pathToContent: string,
  absoluteToRelative: (path: string) => string,
  debug: boolean,
  gitRepo?: GitRepo,
) => {
  if (gitRepo) {
    const { remoteUrl, branch } = gitRepo;
    if (!remoteUrl || !branch) {
      throw Error('[nextmd] You must specify both git remote URL and branch when using `contentGitRepo`');
    }

    /**
     * @returns The number of seconds since the last update of the git remote repo.
     */
    const elapsedSecondsSinceLastUpdate = () => {
      const lastRepoUpdateTxt = fs.readFileSync(pathToNextmdLastUpdate).toString('utf-8').trim();
      const lastUpdateMillis = parseInt(lastRepoUpdateTxt, 10);

      return Date.now() - lastUpdateMillis;
    };

    /**
     *
     * @returns The branche name of the last git clone
     */
    const branchLastUpdate = () => fs.readFileSync(pathToNextmdBranch).toString('utf-8').trim();

    /**
     * Mechanism to avoid pulling the repo when `getStaticPaths` & `getStaticProps` is called.
     *
     * This ensures repo content is the same in `getStaticPaths` & `getStaticProps`.
     */
    const shouldUpdateGitRepo = () => {
      if (fs.existsSync(pathToNextmdLastUpdate) === false || fs.existsSync(pathToNextmdBranch) === false) {
        return 'first time fetch';
      }

      const lastBranch = branchLastUpdate();
      if (lastBranch !== branch) {
        return ['was', lastBranch].join(' ');
      }

      if (elapsedSecondsSinceLastUpdate() > 5 * 60 * 1000) {
        // 5 minutes
        return 'last update was > 5 min ago';
      }

      return null;
    };

    const shouldUpdateGitRepoReason = shouldUpdateGitRepo();
    const logFromGit = [remoteUrl, `(Branch: ${branch})`].filter((e) => e).join(' ');
    if (shouldUpdateGitRepoReason) {
      consoleLogNextmd(
        ['cloning', logFromGit, debug ? `- ${shouldUpdateGitRepoReason}` : undefined].filter((e) => e).join(' '),
      );
      fs.rmSync(pathToContent, { recursive: true, force: true });
      await cmd(
        ['git', 'clone', branch ? `-b ${branch}` : undefined, '--depth 1', remoteUrl, pathToContent]
          .filter((e) => e)
          .join(' '),
      );
      fs.writeFileSync(pathToNextmdBranch, await cmd(`git -C ${pathToContent} rev-parse --abbrev-ref HEAD`));
      fs.writeFileSync(pathToNextmdLastUpdate, `${Date.now()}`);
    }

    consoleLogNextmd('creating page from', logFromGit);
  } else {
    consoleLogNextmd('creating page from', pathToContent);
  }

  return treeSync(pathToContent, absoluteToRelative);
};
