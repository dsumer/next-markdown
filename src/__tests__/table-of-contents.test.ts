import { getTableOfContents } from '../utils/table-of-contents';

describe('Table of Contents', () => {
  test('is created correctly', () => {
    const content = `
# Title 1
## Title 2
## Title 3
### Title 4
#### Title 5
### Title 6
## Title 7
### Title 8
# Title 9
abc`.trim();

    expect(getTableOfContents(content)).toMatchSnapshot();
  });

  test('is created correctly without H1', () => {
    const content = `
## Title 1
#### Title 2
### Title 3
## Title 4
### Title 5
abc`.trim();

    expect(getTableOfContents(content)).toMatchSnapshot();
  });
});
