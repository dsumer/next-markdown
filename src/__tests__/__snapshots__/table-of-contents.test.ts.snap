// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table of Contents is created correctly 1`] = `
Array [
  Object {
    "id": "title-1",
    "level": 1,
    "subItems": Array [
      Object {
        "id": "title-2",
        "level": 2,
        "subItems": Array [],
        "text": "Title 2",
      },
      Object {
        "id": "title-3",
        "level": 2,
        "subItems": Array [
          Object {
            "id": "title-4",
            "level": 3,
            "subItems": Array [
              Object {
                "id": "title-5",
                "level": 4,
                "subItems": Array [],
                "text": "Title 5",
              },
            ],
            "text": "Title 4",
          },
          Object {
            "id": "title-6",
            "level": 3,
            "subItems": Array [],
            "text": "Title 6",
          },
        ],
        "text": "Title 3",
      },
      Object {
        "id": "title-7",
        "level": 2,
        "subItems": Array [
          Object {
            "id": "title-8",
            "level": 3,
            "subItems": Array [],
            "text": "Title 8",
          },
        ],
        "text": "Title 7",
      },
    ],
    "text": "Title 1",
  },
  Object {
    "id": "title-9",
    "level": 1,
    "subItems": Array [],
    "text": "Title 9",
  },
]
`;

exports[`Table of Contents is created correctly without H1 1`] = `
Array [
  Object {
    "id": "title-1",
    "level": 2,
    "subItems": Array [
      Object {
        "id": "title-2",
        "level": 4,
        "subItems": Array [],
        "text": "Title 2",
      },
      Object {
        "id": "title-3",
        "level": 3,
        "subItems": Array [],
        "text": "Title 3",
      },
    ],
    "text": "Title 1",
  },
  Object {
    "id": "title-4",
    "level": 2,
    "subItems": Array [
      Object {
        "id": "title-5",
        "level": 3,
        "subItems": Array [],
        "text": "Title 5",
      },
    ],
    "text": "Title 4",
  },
]
`;
